python/kserve/kserve/model.py

```diff
@@ -218,23 +222,36 @@ async def preprocess(self, payload: Union[Dict, CloudEvent, ModelInferRequest],
 
         return response
 
-    def postprocess(self, response: Union[Dict, ModelInferResponse], headers: Dict[str, str] = None) -> Dict:
-        """The postprocess handler can be overridden for inference response transformation
+    def postprocess(self, response: Union[Dict, InferResponse, ModelInferResponse], headers: Dict[str, str] = None) \
+            -> Union[Dict, ModelInferResponse]:
+        """The postprocess handler can be overridden for inference response transformation.
+        The default implementation converts the v2 infer response types to gRPC or REST.
+        For gRPC request it converts InferResponse to gRPC message or directly returns ModelInferResponse from
+        predictor call.
+        For REST request it converts ModelInferResponse to Dict or directly returns from predictor call.
 
         Args:
-            response (Dict|ModelInferResponse): The response passed from ``predict`` handler.
+            response (Dict|InferResponse|ModelInferResponse): The response passed from ``predict`` handler.
+            headers (Dict): Request headers.
 
         Returns:
             Dict: post-processed response.
         """
         if headers:
-            if "grpc" in headers.get("user-agent", "") and isinstance(response, ModelInferResponse):
-                return response
-            if "application/json" in headers.get("content-type", "") and isinstance(response, ModelInferResponse):
-                return convert_grpc_response_to_dict(response)
+            if "grpc" in headers.get("user-agent", ""):
+                if isinstance(response, ModelInferResponse):
+                    return response
+                elif isinstance(response, InferResponse):
```

Comment:
predict handler can return the `InferResponse` and the postprocess default handler converts to gRPC or REST based on the request type.